rate_limit.py

```diff
@@ -0,0 +1,103 @@
+"""
+This module contains functions for rate limiting requests.
+"""
+
+from datetime import datetime
+import signal
+import sys
+from typing import Dict
+from uuid import uuid4
+
+from apscheduler.schedulers import background
+import gradio as gr
+
+
+class InvalidTokenException(Exception):
+  pass
+
+
+class UserRateLimitException(Exception):
+  pass
+
+
+class SystemRateLimitException(Exception):
+  pass
+
+
+class RateLimiter:
+
+  def __init__(self, limit=10000, period_in_seconds=60 * 60 * 24):
+    # Maps tokens to the last time they made a request.
+    # E.g, {"sometoken": datetime(2021, 8, 1, 0, 0, 0)}
+    self.last_request_times: Dict[str, datetime] = {}
+
+    # The number of requests made.
+    # This count is reset to zero at the end of each period.
+    self.request_count = 0
+
+    # The maximum number of requests allowed within the time period.
+    self.limit = limit
+
+    self.scheduler = background.BackgroundScheduler()
+    self.scheduler.add_job(self.clean_up, "interval", seconds=60 * 60 * 24)
+    self.scheduler.add_job(self.reset_request_count,
+                           "interval",
+                           seconds=period_in_seconds)
+    self.scheduler.start()
+
+  def request_allowed(self, token: str):
+    if not token or token not in self.last_request_times:
+      raise InvalidTokenException()
+
+    if (datetime.now() - self.last_request_times[token]).seconds < 5:
+      raise UserRateLimitException()
+
+    if self.request_count >= self.limit:
+      raise SystemRateLimitException()
+
+    self.last_request_times[token] = datetime.now()
+    self.request_count += 1
+
+  def initialize_request(self, token: str):
+    self.last_request_times[token] = datetime.min
+
+  def clean_up(self):
```

Comment:
`remove_old_tokens`?

Reply:
maybe lower score cuz it's not public?
https://google.github.io/styleguide/pyguide.html#3162-naming-conventions